lib/client/api.go

```diff
@@ -681,74 +696,95 @@ func (tc *TeleportClient) runShell(nodeClient *NodeClient, sessionID session.ID,
 	if tc.Stderr == nil {
 		tc.Stderr = os.Stderr
 	}
+	attachedTerm := (stdin == os.Stdin && term.IsTerminal(0))
+
+	winSize := &term.Winsize{Width: 80, Height: 25}
+	if attachedTerm {
+		winSize, err = term.GetWinsize(0)
+		if err != nil {
+			log.Error(err)
+		}
+	}
+
+	shell, err := nodeClient.Shell(
+		int(winSize.Width),
+		int(winSize.Height),
+		sessionID,
+		tc.Config.Env,
+		attachedTerm)
+
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	defer shell.Close()
+
+	// user-supplied callback
+	if tc.OnShellCreated != nil {
+		exit, err := tc.OnShellCreated(shell)
+		if exit {
+			return err
```

Comment:
also can add optional user friendly description, e.g. `trace.Wrap(err, "failed to call user callback")` or something like this